homeassistant/components/thethingsnetwork/entity.py

```diff
@@ -0,0 +1,203 @@
+"""Support for The Things Network entities."""
+
+from abc import ABC, abstractmethod
+import logging
+from typing import TYPE_CHECKING, Optional
+
+from ttn_client import TTNBaseValue, TTNSensorValue
+
+from homeassistant.core import callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+if TYPE_CHECKING:
+    from homeassistant.config_entries import ConfigEntry
+
+    from .coordinator import TTNCoordinator
+
+from .const import (
+    CONF_APP_ID,
+    OPTIONS_DEVICE_NAME,
+    OPTIONS_FIELD_ICON,
+    OPTIONS_FIELD_NAME,
+    OPTIONS_FIELD_PICTURE,
+    OPTIONS_FIELD_UNIT_MEASUREMENT,
+)
+from .entry_settings import TTN_EntrySettings
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TTN_Entity(CoordinatorEntity, Entity, ABC):
```

Comment:
```suggestion
class TTN_Entity(CoordinatorEntity[...]):
```
Please add the coordinator type as type generic to the coordinator entity